designs/2024-per-rule-autofix-configuration/README.md

```diff
@@ -0,0 +1,406 @@
+
+```bash
+npx eslint --fix --rule 'no-var: { autofix: false }' 
+```
+
+We extend directive parser engine to support new format as well:
+
+```js
+/* eslint eqeqeq: "off", curly: { severity: "warn", autofix: false } */
+```
+
+### Example
+
+```js
+export default [
+  {
+    rules: {
+      // Full rule entry with disabled autofix
+      'eqeqeq': {
+        severity: 'error',
+        options: ['always'],
+        autofix: false,
+      },
+      // Options can be omitted if not needed
+      'no-var': {
+        severity: 'error',
+        autofix: false,
+      },
+      // Severity can be omitted if not needed
+      'camelcase': {
+        options: [{ properties: 'never' }],
+        autofix: false,
+      },
+      // Autofix can be omitted if not needed
+      'yoda': {
+        options: ['never'],
+      },
+      // Old format remains valid and will be normalized internally
+      'no-regex-spaces': 'error',
+      'no-unneeded-ternary': ['error', {
+        defaultAssignment: false,
+      }]
+    },
+  },
+];
+```
+
+> [!NOTE]
+> In practice, users only need the full form when disabling autofix 
+> or when being explicitly descriptive.
+
+<details>
+  <summary>Backward compatibility: existing configurations</summary>
+  
+  ```js
+  export default [
+    {
+      linterOptions: {
+        reportUnusedDisableDirectives: 'error',
+      },
+      rules: {
+        'eqeqeq': 'error',
+        'no-var': ['warn', 'always'],
+        '@scope/some-rule': ['warn', 'always', { option: true }],
+      },
+    },
+  ];
+  ```
+
+  Will be normalized as:
+
+  ```js
+  export default [
+    {
+      linterOptions: {
+        reportUnusedDisableDirectives: {
+          severity: 'error',
+          autofix: true,
+        },
+      },
+      rules: {
+        'eqeqeq': {
+          severity: 'error',
+          options: [],
+          autofix: true,
+        },
+        'no-var': {
+          severity: 'warn',
+          options: ['always'],
+          autofix: true,
+        },
+        'some-plugin/some-rule': {
+          severity: 'warn',
+          options: ['always', { option: true }],
+          autofix: true,
+        },
+      },
+    },
+  ];
+  ```
+</details>
+
+### Implementation plan
+
+#### 1. Normalize rule entries and configuration objects to the extended format
+
+> [!NOTE]
+> This is actually not fully related to the problem of this RFC, just a step to implementation and future extensibility.
+
+As a first step, we normalize rule entries and `linterOptions.reportUnusedDisableDirectives` to the new extended format.
+
+Relevant starting points:
+
+* [@eslint/core > types.ts > LinterOptionsConfig](https://github.com/eslint/rewrite/blob/2020d38f9dbaabb9923b8f2116e7bcbafa530c85/packages/core/src/types.ts#L659)
+* [@eslint/core > types.ts > RuleConfig](https://github.com/eslint/rewrite/blob/2020d38f9dbaabb9923b8f2116e7bcbafa530c85/packages/core/src/types.ts#L679)
```

Comment:
Can we add https://github.com/eslint/rewrite/blob/88525272b717efea7cae1d5ea2429d6343a7e066/packages/plugin-kit/src/config-comment-parser.js#L30-L37 to the list? So we don't forget to check it.